File: Lab/lab6/lab6.py
def canFinish(numCourses, prerequisites):
    # Create a graph from the prerequisites
    graph = {i: [] for i in range(numCourses)}
    for course, prereq in prerequisites:
        graph[course].append(prereq)

    # Keep track of visited nodes to detect cycles
    visited = [0] * numCourses

    def dfs(course):
        # If the node is visited, a cycle is found
        if visited[course] == -1:
            return False
        # If the node is processed, no cycle is found
        if visited[course] == 1:
            return True
        
        # Mark as visited
        visited[course] = -1
        # Check all adjacent vertices
        for prereq in graph[course]:
            if not dfs(prereq):
                return False
        
        # Mark as processed
        visited[course] = 1
        return True
    
    # Check each course for cycles
    for course in range(numCourses):
        if not dfs(course):
            return False
    
    return True

# Test cases
print("Test Case 1:", canFinish(2, [[1,0]]))  # Expected: True
print("Test Case 2:", canFinish(2, [[1,0], [0,1]]))  # Expected: False

# Add more test cases if needed

File: Lab/lab6/lab6.java
import java.util.*;

public class lab6{
    public static boolean canFinish(int numCourses, int[][] prerequisites) {
        ArrayList<ArrayList<Integer>> graph = new ArrayList<>();
        for (int i = 0; i < numCourses; i++) {
            graph.add(new ArrayList<>());
        }

        for (int[] pair : prerequisites) {
            graph.get(pair[0]).add(pair[1]);
        }

        boolean[] visited = new boolean[numCourses];
        boolean[] recStack = new boolean[numCourses];

        for (int course = 0; course < numCourses; course++) {
            if (!dfs(course, graph, visited, recStack)) {
                return false;
            }
        }

        return true;
    }

    private static boolean dfs(int course, ArrayList<ArrayList<Integer>> graph, boolean[] visited, boolean[] recStack) {
        if (recStack[course]) {
            return false;
        }

        if (visited[course]) {
            return true;
        }

        visited[course] = true;
        recStack[course] = true;

        List<Integer> children = graph.get(course);
        for (Integer child : children) {
            if (!dfs(child, graph, visited, recStack)) {
                return false;
            }
        }

        recStack[course] = false;
        return true;
    }

    public static void main(String[] args) {
        // Test Case 1
        int numCourses1 = 2;
        int[][] prerequisites1 = {{1, 0}};
        System.out.println("Test Case 1: " + canFinish(numCourses1, prerequisites1));  // Expected: true

        // Test Case 2
        int numCourses2 = 2;
        int[][] prerequisites2 = {{1, 0}, {0, 1}};
        System.out.println("Test Case 2: " + canFinish(numCourses2, prerequisites2));  // Expected: false
    }
}
